src/main/cpp/src/NativeParquetJni.cpp

```diff
@@ -106,215 +125,272 @@ class column_pruner {
      */
     column_pruner(const std::vector<std::string> & names, 
             const std::vector<int> & num_children, 
-            int parent_num_children): children(), s_id(0), c_id(-1) {
-      add_depth_first(names, num_children, parent_num_children);
+            const std::vector<int> & tags, 
+            int parent_num_children): children(), tag(STRUCT_TAG) {
+      add_depth_first(names, num_children, tags, parent_num_children);
+    }
+
+    column_pruner(int in_tag): children(), tag(in_tag) {
+    }
+
+    column_pruner(): children(), tag(STRUCT_TAG) {
+    }
+
+    void skip(std::vector<parquet::format::SchemaElement> & schema,
+            std::size_t & current_input_schema_index, std::size_t & next_input_chunk_index) {
+      // We want to skip everything referenced by the current_input_schema_index and its children.
+      // But we do have to update the chunk indexes as we go.
+      int num_to_skip = 1;
+      while(num_to_skip > 0 && current_input_schema_index < schema.size()) {
+        auto schema_item = schema[current_input_schema_index];
+        bool is_leaf = schema_item.__isset.type;
+        if (is_leaf) {
+          ++next_input_chunk_index;
+        }
+
+        if (schema_item.__isset.num_children) {
+          num_to_skip = num_to_skip + schema_item.num_children;
+        }
+
+        --num_to_skip;
+        ++current_input_schema_index;
+      }
     }
 
-    column_pruner(int s_id, int c_id): children(), s_id(s_id), c_id(c_id) {
+    std::string get_name(parquet::format::SchemaElement & elem, const bool normalize_case = false) {
+      if (normalize_case) {
+        return unicode_to_lower(elem.name);
+      } else {
+        return elem.name;
+      }
     }
 
-    column_pruner(): children(), s_id(0), c_id(-1) {
+    column_pruner * find_child(std::string name) {
+      auto found_it = children.find(name);
+      if (found_it != children.end()) {
+        return &(found_it->second);
+      }
+      return nullptr;
```

Comment:
Yes, but I am not sure how to do it with the iterators in a clean way. Suggestions are welcome.

Reply:
The solution here would be returning a reference instead of a pointer. But by doing so, this function must throw a not found exception if the element doesn't exist. As such, you can do something like:
```
const column_pruner& find_child(std::string name) {
      auto found_it = children.find(name);
      if (found_it != children.end()) {
        throw exception(...);
      }
      return found_it->second;
}

if(exist_child(name)) {
  auto const& child = find_child(name);
  // process child
}

```

Reply:
my first thought was to simply return the iterator, but then you have the issue of the caller having to know the .end() of children, which is an implementation detail. I'm leaning towards an optional string return value right now.

Reply:
std::optional does not work for references. Apparently it is a union and you cannot store a reference inside a union. At least that is the error message I got when I tried to use it.

Reply:
Yes, in order to use a reference in `<optional>` you need to wrap it in `reference_wrapper`. Here is an example: https://godbolt.org/z/o3MhMfczY